# Obtain a reference genome
either from Ensembl, iGenomes, NCBI or UCSC.

In this course, we will use the human GRCh38 version of the genome from Ensembl. Furthermore, we are actually going to perform the analysis using only a single chromosome (chr22) and the ERCC spike-in to make it run faster. Griffith lab have prepared this simplified reference for you. It contains chr22 (and ERCC transcript) fasta files in both a single combined file and individual files. 

In [1]:
# Our current directory when you first SSH into EC2 instance:
pwd

/home/ubuntu


In [2]:
ls

Miniconda3-latest-Linux-x86_64.sh  R           workspace
Module00_install_bedtools.ipynb    bin
Module01_referenceGenomes.ipynb    miniconda3


Earlier we installed tools to $RNA_HOME/student_tools directory
Today, we'll install reference genome to $RNA_REFS_DIR directory
Knowing that:
RNA_HOME=/home/ubuntu/workspace/rnaseq
RNA_REFS_DIR=/home/ubuntu/workspace/rnaseq/refs

## Make required directory to install refernce genome file into
If you just finished intalling tools and immediately move on to this step, you can go back to $RNA_HOME by:

In [3]:
cd $RNA_HOME
# Note that $RNA_HOME=/home/ubuntu/workspace/rnaseq
# Curently, $RNA_HOME directory only contains student_tools folder/directory because we created this directory to install our tools earlier
ls

student_tools


In [4]:
# We already create RNA_REFS_DIR=/home/ubuntu/workspace/rnaseq/refs in our .bashrc file earlier, just to confirm that the path is set correctly:
echo $RNA_REFS_DIR

/home/ubuntu/workspace/rnaseq/refs


In [5]:
# Now that we have the path set correctly, let's actually create this directory:
mkdir -p $RNA_REFS_DIR

In [6]:
# You're currently in $RNA_HOME=/home/ubuntu/workspace/rnaseq, if we ls the content of it again, we'll now see "refs" directory
ls

refs  student_tools


## Install the reference genome fasta file

In [7]:
# Let's go into $RNA_REFS_DIR and install the ref genome fasta file there 
cd $RNA_REFS_DIR # or cd refs
wget http://genomedata.org/rnaseq-tutorial/fasta/GRCh38/chr22_with_ERCC92.fa
ls # you should see the chr22_with_ERCC92.fa file  

--2025-04-26 15:39:24--  http://genomedata.org/rnaseq-tutorial/fasta/GRCh38/chr22_with_ERCC92.fa
Resolving genomedata.org (genomedata.org)... 54.71.55.4
Connecting to genomedata.org (genomedata.org)|54.71.55.4|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://genomedata.org/rnaseq-tutorial/fasta/GRCh38/chr22_with_ERCC92.fa [following]
--2025-04-26 15:39:25--  https://genomedata.org/rnaseq-tutorial/fasta/GRCh38/chr22_with_ERCC92.fa
Connecting to genomedata.org (genomedata.org)|54.71.55.4|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 51751056 (49M)
Saving to: ‘chr22_with_ERCC92.fa’

chr22_with_ERCC92.f 100%[===================>]  49.35M  32.2MB/s    in 1.5s    

2025-04-26 15:39:26 (32.2 MB/s) - ‘chr22_with_ERCC92.fa’ saved [51751056/51751056]

chr22_with_ERCC92.fa


## Examine reference genome files

In [13]:
# View a list of all sequences (all header lines) in our reference genome fasta file:
grep ">" chr22_with_ERCC92.fa
# You'll see chr22 header line, and ERCC header lines indicating this files contains chr22 and ERCC transcripts

>22 dna_sm:chromosome chromosome:GRCh38:22:1:50818468:1 REF
>ERCC-00002
>ERCC-00003
>ERCC-00004
>ERCC-00009
>ERCC-00012
>ERCC-00013
>ERCC-00014
>ERCC-00016
>ERCC-00017
>ERCC-00019
>ERCC-00022
>ERCC-00024
>ERCC-00025
>ERCC-00028
>ERCC-00031
>ERCC-00033
>ERCC-00034
>ERCC-00035
>ERCC-00039
>ERCC-00040
>ERCC-00041
>ERCC-00042
>ERCC-00043
>ERCC-00044
>ERCC-00046
>ERCC-00048
>ERCC-00051
>ERCC-00053
>ERCC-00054
>ERCC-00057
>ERCC-00058
>ERCC-00059
>ERCC-00060
>ERCC-00061
>ERCC-00062
>ERCC-00067
>ERCC-00069
>ERCC-00071
>ERCC-00073
>ERCC-00074
>ERCC-00075
>ERCC-00076
>ERCC-00077
>ERCC-00078
>ERCC-00079
>ERCC-00081
>ERCC-00083
>ERCC-00084
>ERCC-00085
>ERCC-00086
>ERCC-00092
>ERCC-00095
>ERCC-00096
>ERCC-00097
>ERCC-00098
>ERCC-00099
>ERCC-00104
>ERCC-00108
>ERCC-00109
>ERCC-00111
>ERCC-00112
>ERCC-00113
>ERCC-00116
>ERCC-00117
>ERCC-00120
>ERCC-00123
>ERCC-00126
>ERCC-00130
>ERCC-00131
>ERCC-00134
>ERCC-00136
>ERCC-00137
>ERCC-00138
>ERCC-00142
>ERCC-00143
>ERCC-00144
>ERCC-00145
>ERCC-00147
>ERC

In [9]:
# View the first 10 lines of this file
head chr22_with_ERCC92.fa

>22 dna_sm:chromosome chromosome:GRCh38:22:1:50818468:1 REF
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN


### Estimated length of chr22 + ERCC in base pairs
The reference human chromosome 22 is around 50.8 million base pairs long (GRCh38), so most of this file is likely chr22, and only a small part (less than 1 Mbp) is from the ERCC transcript sequence.

By default, wc shows three counts for a file or input:
* Number of lines ("wc -l" will output only number of lines)
* Number of words
* Number of bytes or How long is this chromosome (in bases and Mbp). ~51,751,056 bases or ~51.75 Mbp (assuming 1 character A/T/C/G/N = 1 base)
* the name of the file

In [10]:
wc chr22_with_ERCC92.fa

  848761   848764 51751056 chr22_with_ERCC92.fa


In [11]:
# Because chr22_with_ERCC92.fa contains 848764 lines, to view 10 lines from approximately the middle of this file:
head -n 425000 chr22_with_ERCC92.fa | tail # first print the first 425000 of chr22_with_ERCC92.fa, then view only the last 10 lines of these lines

ggaggctgaggcaggagaatcgcttgaacatgggaggtggaagttgcagtgagccgaaac
tgcgccattgcactatagcctgggcaacaagagtgaaagtctgtcttgaaaaaaaaaaaT
CAGATGTTCTATGTAAAAATGCTATCTAtgattgaagtataaaactttacctccctttat
gttcctttgccctccccactatttattattgtcttgattatatcttctatatgcattgag
aggtgttataacttttgtatcaatcaccaaatttaatttagaaaatataagaggagaaga
aaagtctattacatttactcatatttttgcttactgtgttctttcttccttcttgatgtt
ccagaatttcttttattgcttcttttctgcttagaaaactttatctttttctttcatctt
tcttttttcctcctcctcctcctcctcctttttttttttttttttttttttttttttaat
aaagagacagggtctcactctatcacccagactggagttcagtgatgcaatcatagctca
ttgcaaccttgaactcctgggctcaagtgatcctcccacctcagcctcctgagtagctgg


In [12]:
# What is the count of each base in the entire reference genome file (skipping the header lines for each sequence)?
grep -v ">" chr22_with_ERCC92.fa | perl -ne 'chomp $_; $bases{$_}++ for split //; if (eof){print "$_ $bases{$_}\n" for sort keys %bases}'
# grep -v ">" remove lines containing ">". These are FASTA header lines
# | is piping command, it takes the output of preceeding command as input for the next command
# perl -n loops over each input line without automatically printing anything
# split // splits the line into individual characters (i.e., bases)
# For each character, increment its count in the %bases hash
# eof is true only at the end of the last line of input
## When we reach the end of the file:
### Loop over the sorted keys (base letters like A, C, G, T, N, etc.)
### Print each base and its count, one per line.

A 4455938
C 4406493
G 4411768
N 10710000
T 4445994
Y 1
a 5950524
c 4772185
g 4853055
n 948691
t 5946575


### How many occurences of the EcoRI restriction site are present in the chromosome 22 sequence? 
The EcoRI restriction enzyme recognition sequence is 5’-GAATTC-‘3. Since this is a palendrome, the reverse complement is the same and we only have to search for one sequence in our string. After accounting for end of line breaks and case sensitivity we find 3,935 occurences of this sequence.

In [14]:
# To extract only the chr22 sequence into a new FASTA file called chr22_only.fa from chr22_with_ERCC92.fa
awk '/^>/ {keep = /^>22 /} keep' chr22_with_ERCC92.fa > chr22_only.fa

* wk: processes the FASTA file line by line.
* /^>/: matches header lines.
* keep = /^>22 /: sets a flag (keep) to true only for the chr22 header line (>22 ...).
* keep: if the flag is true, the line (header or sequence) is printed.

It will stop copying when it reaches the next > line (e.g., >ERCC-00002)

In [16]:
# Check if I extracted chr22 correctly
grep ">" chr22_only.fa
head -n 3 chr22_only.fa
tail -n 2 chr22_only.fa

>22 dna_sm:chromosome chromosome:GRCh38:22:1:50818468:1 REF
>22 dna_sm:chromosome chromosome:GRCh38:22:1:50818468:1 REF
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
nnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn
nnnnnnnnnnnnnnnnnnnnnnnnnnnn


#### My method (ChatGPT's method)
* Remove FASTA headers (start with >)
* Joins all sequence lines into one continuous string
* Uppercases all letters
* Counts all non-overlapping GAATTC matches

In [17]:
# Count occurence of EcoRI restriction site, and time how long it takes to generate output:
time (grep -v "^>" chr22_only.fa | tr -d '\n' | tr 'a-z' 'A-Z' | grep -o 'GAATTC' | wc -l)

# tr -d '\n' removes line breaks, so sequences are treated as one long string
# tr 'a-z' 'A-Z' converts lowercase to uppercase (makes search case-insensitive)
# grep -o "GAATTC" finds and prints each match of the EcoRI site
# wc -l counts how many times it was matched (i.e., how many times GAATTC appears)

9544

real	0m0.924s
user	0m1.029s
sys	0m0.184s


#### Instructor's method
* grep -v ">": removes headers ✅
* First Perl: uppercases each line, removes trailing newline, but prints each line separately
* Second Perl: processes each line one at a time, counts how many GAATTC matches exist per line

❗ Problem:
If a GAATTC recognition site is split across two lines, like this:

...GAATT

C...

the instructor's method will not count it, because it's analyzing each line independently, not a continuous string.

In [18]:
# Instructor's method to count occurrence of EcoRI restriction site
time (grep -v ">" chr22_only.fa | 
perl -ne 'chomp $_; $s = uc($_); print $_;' | 
perl -ne '$c += $_ =~ s/GAATTC/XXXXXX/g; if (eof){print "\nEcoRI site (GAATTC) count = $c\n\n";}')


EcoRI site (GAATTC) count = 3935


real	0m0.274s
user	0m0.269s
sys	0m0.147s


#### Conlusion: I think ChatGPT's code is more accurate. Instructor's method undercounts the occurrence of EcoRI restriction site